azure-ts-dynamicresource/cdnCustomDomain.ts

```diff
@@ -0,0 +1,276 @@
+            result = {
+                failures: [
+                    {
+                        property: this.name,
+                        reason: validationOutput.reason || "domain_validation_failed"
+                    }
+                ]
+            };
+        } else {
+            result = {
+                inputs: news
+            };
+        }
+
+        return result;
+    }
+
+    async diff(id: string, previousOutput: DynamicProviderOutputs, news: DynamicProviderInputs): Promise<pulumi.dynamic.DiffResult> {
+        const replaces: string[] = [];
+        let changes = false;
+        let deleteBeforeReplace = false;
+
+        if (previousOutput.customDomainHostName !== news.customDomainHostName || previousOutput.name !== this.name) {
+            console.warn("Changing the domain name properties will cause a downtime.")
+
+            deleteBeforeReplace = true;
+            if (previousOutput.customDomainHostName !== news.customDomainHostName) {
+                replaces.push("customDomainHostName");
+            }
+            if (previousOutput.name !== this.name) {
+                replaces.push("customDomainName");
+            }
+        }
+
+        if (previousOutput.endpointName !== news.endpointName) {
+            deleteBeforeReplace = true;
+            replaces.push("endpointName");
+        }
+
+        // HTTPS can be enabled/disabled in-place.
+        if (previousOutput.httpsEnabled !== news.httpsEnabled) {
+            changes = true;
+            replaces.push("httpsEnabled");
+        }
+
+        return {
+            deleteBeforeReplace: deleteBeforeReplace,
+            replaces: replaces,
+            changes: changes
+        };
+    }
+
+    async create(inputs: DynamicProviderInputs): Promise<pulumi.dynamic.CreateResult> {
+        const cdnClient = await this.getCDNManagementClient();
+        const result =
+            await cdnClient.customDomains
+                .create(
+                    inputs.resourceGroupName,
+                    inputs.profileName,
+                    inputs.endpointName,
+                    this.name,
+                    inputs.customDomainHostName);
+
+        if (inputs.httpsEnabled) {
+            console.log("Enabling HTTPS for the custom domain");
+
+            await cdnClient.customDomains
+                .enableCustomHttps(
+                    inputs.resourceGroupName,
+                    inputs.profileName,
+                    inputs.endpointName,
+                    this.name);
+        }
+
+        const outs: DynamicProviderOutputs = {
+            ...result,
+            name: this.name,
+            resourceGroupName: inputs.resourceGroupName,
+            profileName: inputs.profileName,
+            endpointName: inputs.endpointName,
+            customDomainHostName: inputs.customDomainHostName,
+            httpsEnabled: inputs.httpsEnabled,
+        };
+
+        return {
+            id: result.id!,
```

Comment:
not clear why the `!` is necessary.

Reply:
It is an optional in the `result`. I get a lint error for not checking for null-ness otherwise. Actually, I have a question for @pgavlin here. Should I be setting the value of the `id` in the output returned by the `create()`?

Reply:
> Should I be setting the value of the id in the output returned by the create()?

Yes - you'll want to ensure you return an `id` that represents the resource in the backing cloud provider.  This may be necessary for example to `delete` the resource later - as you'll need to use this `id` to look it up.

It's slightly concerning that the `id` is an optional return value from the cloud provider!

Reply:
> It's slightly concerning that the id is an optional return value from the cloud provider!

Yes, agreed. Though, it seems that the custom domain resource does always get an ID. I believe this is an optional because the underlying CDN management client extends itself from another type `Resource`, which is where this `id` comes from and it would seem that it is optional for other resources. So I think we are safe in this instance.